pkg/arvo/lib/group-store.hoon

```diff
@@ -0,0 +1,297 @@
+      %ban-ships  [ranks+(set ship ships.diff) ~]
+    ==
+  ::
+  ++  remove-group
+    |=  =^update
+    ^-  json
+    ?>  ?=(%remove-group -.update)
+    (frond %group-id (group-id group-id.update))
+  ::
+  ++  group-id
+    |=  =^group-id
+    ^-  json
+    %-  pairs
+    :~  name+s+term.group-id
+        ship+(ship ship.group-id)
+    ==
+  ++  add-group
+    |=  =action
+    ^-  json
+    ?>  ?=(%add-group -.action)
+    %-  pairs
+    :~  group-id+(group-id group-id.action)
+        policy+(policy policy.action)
+    ==
+  ::
+  ++  add-members
+    |=  =action
+    ^-  json
+    ?>  ?=(%add-members -.action)
+    %-  pairs
+    :~  group-id+(group-id group-id.action)
+        ships+(set ship ships.action)
+    ==
+  ::
+  ++  remove-members
+    |=  =action
+    ^-  json
+    ?>  ?=(%remove-members -.action)
+    %-  pairs
+    :~  group-id+(group-id group-id.action)
+        ships+(set ship ships.action)
+    ==
+  ::
+  ++  add-tag
+    |=  =action
+    ^-  json
+    ?>  ?=(%add-tag -.action)
+    %-  pairs
+    :~  group-id+(group-id group-id.action)
+        tag+s+tag.action
+        ships+(set ship ships.action)
+    ==
+  ::
+  ++  remove-tag
+    |=  =action
+    ^-  json
+    ?>  ?=(%remove-tag -.action)
+    %-  pairs
+    :~  group-id+(group-id group-id.action)
+        tag+s+tag.action
+        ships+(set ship ships.action)
+    ==
+  ::
+  ++  change-policy
+    |=  =action
+    ^-  json
+    ?>  ?=(%change-policy -.action)
+    %-   pairs
+    :~   group-id+(group-id group-id.action)
+         diff+(policy-diff diff.action)
+    ==
+  --
+++  dejs
+  =,  dejs:format
+  |%
+  ::
+  ++  action
+    ^-  $-(json ^action)
+    %-  of
+    :~
+      add-group+add-group
+      add-members+add-members
+      remove-members+remove-members
+      add-tag+add-tag
+      remove-tag+remove-tag
+      change-policy+change-policy
+      remove-group+remove-group
+    ==
+  ++  rank
+    |=  =json
+    ^-  ^rank
+    ?>  ?=(%n -.json)
+    ?:  =('0' p.json)  %0
+    ?:  =('1' p.json)  %1
+    ?:  =('2' p.json)  %2
+    ?:  =('3' p.json)  %3
+    !!
+  ::  move to zuse also
+  ++  oj
+    |*  =fist
+    ^-  $-(json (jug cord _(fist *json)))
+    (om (as fist))
+  ++  tag-queries
+    ^-  $-(json ^tag-queries)
+    (oj ship)
+  :: TODO: move to zuse
+  ++  ship
+    (su ;~(pfix sig fed:ag))
+  ++  policy
+    ^-  $-(json ^policy)
+    %-  of
+    :~  invite+invite-policy
+        open+open-policy
+    ==
+  ++  invite-policy
+    %-  ot
+    :~  pending+(as ship)
+    ==
+  ++  open-policy
+    %-  ot
+    :~  ranks+(as rank)
+        banned+(as ship)
+    ==
+  ++  policy-diff
+    ^-  $-(json diff:^policy)
+    %-  of
+    :~  add-invites+(as ship)
+        remove-invites+(as ship)
+        allow-ranks+(as rank)
+        allow-ships+(as ship)
+        ban-ranks+(as rank)
+        ban-ships+(as ship)
+        replace+policy
+    ==
+  ++  group-id
+    %-  ot
+    :~  ship+ship
+        name+so
+    ==
+  ::
+  ++  remove-group
```

Comment:
is there a nicer way to do this? `%remove-group` needs the `~` at the end to make sure the axes match so we can always access group-id when given a `action:group-store`

Reply:
One option is to transpose the `group-id` and the tag.  Thus, it's `$:  =group-id  $%  [%remove-group ~] 
 ...  ==`.  You still want the `~`, but some prefer this style since it's more explicit that every case needs a `group-id`.  This may violate local conventions, though, which would take precedence.